tests/unit-tests/frontend_wayland/test_desktop_file_manager.cpp

```diff
@@ -0,0 +1,282 @@
+};
+
+}
+
+struct DesktopFileManager : Test
+{
+    std::shared_ptr<InMemoryDesktopFileCache> cache;
+    std::shared_ptr<mf::DesktopFileManager> file_manager;
+    mtd::MockSurface surface;
+    std::shared_ptr<mtd::MockSceneSession> session;
+
+    const char* APPLICATION_ID = "test_app_id";
+    const char* DESKTOP_FILE_APP_ID = "test_app_id.desktop";
+    const int PID = -12345; // Negative so it never accidentally works
+
+    void SetUp() override
+    {
+        session = std::make_shared<mtd::MockSceneSession>();
+        ON_CALL(*session, process_id)
+            .WillByDefault(testing::Invoke([this]() {
+                return PID;
+            }));
+
+        ON_CALL(surface, application_id())
+            .WillByDefault(testing::Invoke([this]() { return APPLICATION_ID; }));
+
+        ON_CALL(surface, session())
+            .WillByDefault(testing::Invoke([this]() {
+                return session;
+            }));
+
+        cache = std::make_shared<InMemoryDesktopFileCache>();
+        file_manager = std::make_shared<mf::DesktopFileManager>(cache);
+    }
+};
+
+TEST_F(DesktopFileManager, can_find_app_when_app_id_matches)
+{
+    auto new_file = std::make_shared<mf::DesktopFile>(DESKTOP_FILE_APP_ID, nullptr, nullptr);
+    cache->files.push_back(new_file);
+    auto app_id = file_manager->resolve_app_id(&surface);
+    EXPECT_THAT(app_id, Eq(new_file->id));
+}
+
+TEST_F(DesktopFileManager, can_find_app_when_wm_class_matches)
+{
+    auto new_file = std::make_shared<mf::DesktopFile>(nullptr, APPLICATION_ID, nullptr);
+    cache->files.push_back(new_file);
+    auto app_id = file_manager->resolve_app_id(&surface);
+    EXPECT_THAT(app_id, Eq(new_file->id));
+}
+
+TEST_F(DesktopFileManager, can_find_app_when_app_id_matches_despite_being_uppercase)
+{
+    auto uppercase_app_id = std::string(APPLICATION_ID);
+    for(char &ch : uppercase_app_id)
+        ch = std::toupper(ch);
+
+    ON_CALL(surface, application_id())
+        .WillByDefault(testing::Invoke([uppercase_app_id]() { return uppercase_app_id; }));
+
+    auto new_file = std::make_shared<mf::DesktopFile>(DESKTOP_FILE_APP_ID, nullptr, nullptr);
+    cache->files.push_back(new_file);
+    auto app_id = file_manager->resolve_app_id(&surface);
+    EXPECT_THAT(app_id, Eq(new_file->id));
+}
+
+TEST_F(DesktopFileManager, when_every_check_fails_then_application_id_is_returned)
+{
+    auto app_id = file_manager->resolve_app_id(&surface);
+    EXPECT_THAT(app_id, Eq(APPLICATION_ID));
+}
+
+TEST_F(DesktopFileManager, can_find_app_when_app_id_matches_with_gnome_prefix)
+{
+    const std::string GNOME_PREFIX = "gnome-";
+    const std::string gnome_prefixed_desktop_file_app_id = GNOME_PREFIX + DESKTOP_FILE_APP_ID;
+    auto new_file = std::make_shared<mf::DesktopFile>(gnome_prefixed_desktop_file_app_id.c_str(), nullptr, nullptr);
+    cache->files.push_back(new_file);
+    auto app_id = file_manager->resolve_app_id(&surface);
+    EXPECT_THAT(app_id, Eq(new_file->id));
+}
```

Comment:
Isn't this duplicated by the parametrised test below? Do I understand that correctly?